File: fs/Store.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace OSRSCache.fs
{
	using IndexType = OSRSCache.IndexType;
	using DiskStorage = OSRSCache.fs.jagex.DiskStorage;


	public sealed class Store : System.IDisposable
	{
		private readonly Storage storage;
		private readonly IList<Index> indexes = new List<Index>();

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public Store(java.io.File folder) throws java.io.IOException
		public Store(string folder)
		{
			storage = new DiskStorage(folder);
			storage.init(this);
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public Store(Storage storage) throws java.io.IOException
		public Store(Storage storage)
		{
			this.storage = storage;

			storage.init(this);
		}

		public Storage Storage
		{
			get
			{
				return storage;
			}
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: @Override public void close() throws java.io.IOException
		public void Dispose()
		{
			storage.close();
		}

		public override int GetHashCode()
		{
			int hash = 5;
			hash = 79 * hash + this.indexes.GetHashCode();
			return hash;
		}

		public override bool Equals(object obj)
		{
			if (obj == null)
			{
				return false;
			}
			if (this.GetType() != obj.GetType())
			{
				return false;
			}
//JAVA TO C# CONVERTER WARNING: The original Java variable was marked 'final':
//ORIGINAL LINE: final Store other = (Store) obj;
			Store other = (Store) obj;
			if (!Object.Equals(this.indexes, other.indexes))
			{
				return false;
			}
			return true;
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public final Index addIndex(int id) throws java.io.FileNotFoundException
		public Index addIndex(int id)
		{
			foreach (Index i in indexes)
			{
				if (i.Id == id)
				{
					throw new System.ArgumentException("index " + id + " already exists");
				}
			}

			Index index = new Index(id);
			this.indexes.Add(index);

			return index;
		}

		public void removeIndex(Index index)
		{
			Debug.Assert(indexes.Contains(index));
			indexes.Remove(index);
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public void load() throws java.io.IOException
		public void load()
		{
			storage.load(this);
		}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in C#:
//ORIGINAL LINE: public void save() throws java.io.IOException
		public void save()
		{
			storage.save(this);
		}

		public IList<Index> Indexes
		{
			get
			{
				return indexes;
			}
		}

		public Index getIndex(IndexType type)
		{
			return findIndex(type.Number);
		}

		public Index findIndex(int id)
		{
			foreach (Index i in indexes)
			{
				if (i.Id == id)
				{
					return i;
				}
			}
			return null;
		}
	}

}

File: src/main/java/net/runelite/cache/fs/Store.cpp
#include "Store.h"

namespace net::runelite::cache::fs
{
	using Closeable = Closeable;
	using File = java::io::File;
	using FileNotFoundException = java::io::FileNotFoundException;
	using IOException = java::io::IOException;
	using ArrayList = java::util::ArrayList;
	using List = java::util::List;
	using Objects = java::util::Objects;
	using IndexType = net::runelite::cache::IndexType;
	using DiskStorage = net::runelite::cache::fs::jagex::DiskStorage;
	using Logger = org::slf4j::Logger;
	using LoggerFactory = org::slf4j::LoggerFactory;
const std::shared_ptr<org::slf4j::Logger> Store::logger = org::slf4j::LoggerFactory::getLogger(Store::class);

	Store::Store(std::shared_ptr<File> folder) : storage(std::make_shared<DiskStorage>(folder))
	{
//JAVA TO C++ CONVERTER TODO TASK: You cannot use 'shared_from_this' in a constructor:
		storage->init(shared_from_this());
	}

	Store::Store(std::shared_ptr<Storage> storage) : storage(storage)
	{

//JAVA TO C++ CONVERTER TODO TASK: You cannot use 'shared_from_this' in a constructor:
		storage->init(shared_from_this());
	}

	std::shared_ptr<Storage> Store::getStorage()
	{
		return storage;
	}

	void Store::close()
	{
		storage->close();
	}

	int Store::hashCode()
	{
		int hash = 5;
		hash = 79 * hash + Objects::hashCode(this->indexes);
		return hash;
	}

	bool Store::equals(std::any obj)
	{
		if (!obj.has_value())
		{
			return false;
		}
		if (getClass() != obj.type())
		{
			return false;
		}
		constexpr std::shared_ptr<Store> other = std::any_cast<std::shared_ptr<Store>>(obj);
		if (!Objects::equals(this->indexes, other->indexes))
		{
			return false;
		}
		return true;
	}

	std::shared_ptr<Index> Store::addIndex(int id)
	{
		for (auto i : indexes)
		{
			if (i->getId() == id)
			{
				throw std::invalid_argument(StringHelper::wstring_to_string(L"index " + std::to_wstring(id) + L" already exists"));
			}
		}

		std::shared_ptr<Index> index = std::make_shared<Index>(id);
		this->indexes.push_back(index);

		return index;
	}

	void Store::removeIndex(std::shared_ptr<Index> index)
	{
		assert(std::find(indexes.begin(), indexes.end(), index) != indexes.end());
//JAVA TO C++ CONVERTER TODO TASK: The Java Collection 'remove(Object)' method is not converted:
		indexes.remove(index);
	}

	void Store::load()
	{
		storage->load(shared_from_this());
	}

	void Store::save()
	{
		storage->save(shared_from_this());
	}

	std::vector<std::shared_ptr<Index>> Store::getIndexes()
	{
		return indexes;
	}

	std::shared_ptr<Index> Store::getIndex(IndexType type)
	{
		return findIndex(type.getNumber());
	}

	std::shared_ptr<Index> Store::findIndex(int id)
	{
		for (auto i : indexes)
		{
			if (i->getId() == id)
			{
				return i;
			}
		}
		return nullptr;
	}
}
